### DISCUSSION TITLE
Uninitialized constant `Devise::SessionsController` when extending controller

### DISCUSSION THREAD

**Discussion by mcfoton** [activeadmin/activeadmin#8160] [Help] [answered]
Issue happens when changing Rails version from 6.x to 7.x
To enable 2FA for admin dashboard we are overriding the SessionsController.

```
# /app/controllers/active_admin/devise/sessions_controller.rb

module ActiveAdmin
  module Devise
    class SessionsController < ::Devise::SessionsController
      before_action :authenticate_with_two_factor, only: :create

      def authenticate_with_two_factor
        # do something 
      end

    end
  end
end

```

Simply having this file produces an error with rails 7.x

```
.../gems/activeadmin-3.1.0/lib/active_admin/devise.rb:54:in `<module:Devise>': uninitialized constant Devise::SessionsController (NameError)

    class SessionsController < ::Devise::SessionsController
                                       ^^^^^^^^^^^^^^^^^^^^

```

Seems to be related to the loading process, but this setup works with rails 6.x and zeitwerk enabled 🤔 

---
Reproduced with blank Rails app

rails 7.1.2
activeadmin 3.1.0
devise 4.9.3

**Comment by mcfoton** [comment]
update:

The custom controller code does register if I
1. Put the code inside an initializer under `/config/initializers` instead of `/app/controllers`
2. Wrap the code with `Rails.application.config.after_initialize do` as suggested [here](https://guides.rubyonrails.org/autoloading_and_reloading_constants.html#autoload-on-boot-only)

Not sure if this a proper solution here or simply a walk around. Also not sure if I am restricted to have the code under `initializers` folder only.

Would appreciate any suggestions.

**Comment by javierjulio** [comment, marked as answer]
A few things to note. I believe the approach has to be `class ActiveAdmin::Devise::SessionsController` instead of a module with Zeitwerk but not sure if that will help much based on the Rails guide. It could be that it occurs because those files are in our `lib` and not in the `app` folder so I wonder if related to that. I've seen past issues bring this up.

All I can do is share what I've done in my app for a long time now and it works fine with Rails 7 and Zeitwerk. Also, I think this would be preferred since you'd want to customize most things like the view too. We take this approach to use Okta for admin login.

In routes, declare the new sessions controller:

```ruby
  devise_for :admin_users, ActiveAdmin::Devise.config.merge(
    controllers: {
      sessions: 'active_admin/sessions',
      # ...
    }
  )
```

Then declare a controller in `app/controllers/active_admin/sessions_controller.rb` as you would normally along with the view in `app/views/active_admin/sessions/new.html.erb` so you can customize.

```ruby
class ActiveAdmin::SessionsController < ActiveAdmin::Devise::SessionsController
  # ...
end
```

From there it should be ok.

**Comment by mcfoton** [comment]
Awesome, thanks for the prompt response. Your suggested solution works well for me.

It seems that simply having _any_ `.rb` file inside `/controllers/active_admin/devise` is throwing off the loader.
Works no problem with files under `/controller/active_admin`, as you suggested.

**Comment by til** [comment]
I had the same problem and found the solution here, thanks. A small hint for those who also come across this – to override only some Devise controllers, make sure to not clear the `controllers` hash of the config, e.g. by using `deep_merge`:

```ruby
  devise_for :admin_users, ActiveAdmin::Devise.config.deep_merge(
    controllers: {
      sessions: 'active_admin/sessions',
      # ...
    }
  )
```
